src/crimson/osd/pg_backend.cc

```diff
@@ -969,57 +983,61 @@ static int do_xattr_cmp_u64(int op, uint64_t lhs, bufferlist& rhs_xattr)
-  const ObjectState& os,
+static PGBackend::cmp_xattr_ierrorator::future<>
+do_cmp_xattr_mode(
   OSDOp& osd_op,
-  object_stat_sum_t& delta_stats) const
+  bufferlist &&xattr)
 {
-  std::string name{"_"};
   auto bp = osd_op.indata.cbegin();
-  bp.copy(osd_op.op.xattr.name_len, name);
- 
-  logger().debug("cmpxattr on obj={} for attr={}", os.oi.soid, name);
-  return getxattr(os.oi.soid, std::move(name)).safe_then_interruptible(
-    [&delta_stats, &osd_op] (auto &&xattr) {
-    int result = 0;
-    auto bp = osd_op.indata.cbegin();
-    bp += osd_op.op.xattr.name_len;
+  bp += osd_op.op.xattr.name_len;
 
-    switch (osd_op.op.xattr.cmp_mode) {
+  switch (osd_op.op.xattr.cmp_mode) {
     case CEPH_OSD_CMPXATTR_MODE_STRING:
-      {
-        string lhs;
-        bp.copy(osd_op.op.xattr.value_len, lhs);
-        string_view rhs(xattr.c_str(), xattr.length());
-        result = do_cmp_xattr(osd_op.op.xattr.cmp_op, lhs, rhs);
-        logger().debug("cmpxattr lhs={}, rhs={}", lhs, rhs);
-      }
+    {
+      string lhs;
+      bp.copy(osd_op.op.xattr.value_len, lhs);
+      string_view rhs(xattr.c_str(), xattr.length());
+      logger().debug("cmpxattr lhs={}, rhs={}", lhs, rhs);
+      return do_cmp_xattr(osd_op.op.xattr.cmp_op, lhs, rhs);
+    }
     break;
     case CEPH_OSD_CMPXATTR_MODE_U64:
-      {
-        uint64_t lhs;
-        try {
-          decode(lhs, bp);
-	} catch (ceph::buffer::error& e) {
-          logger().info("cmp_xattr: buffer error expection");
-          result = -EINVAL;
-          break;
-	}
-        result = do_xattr_cmp_u64(osd_op.op.xattr.cmp_op, lhs, xattr);
+    {
+      uint64_t lhs;
+      try {
+        decode(lhs, bp);
+      } catch (ceph::buffer::error& e) {
+        logger().info("cmp_xattr: buffer error expection");
+        return crimson::ct_error::invarg::make();
       }
+      return do_xattr_cmp_u64(osd_op.op.xattr.cmp_op, lhs, std::move(xattr));
+    }
     break;
     default:
       logger().info("bad cmp mode {}", osd_op.op.xattr.cmp_mode);
-      result = -EINVAL;
-    }
-    if (result == 0) {
-      logger().info("cmp_xattr: comparison returned false");
-      osd_op.rval = -ECANCELED;
-    } else {
-      osd_op.rval = result;
-    }
-    delta_stats.num_rd++;
-    delta_stats.num_rd_kb += shift_round_up(osd_op.op.xattr.value_len, 10);
+      return crimson::ct_error::invarg::make();
+  }
+}
+
+PGBackend::cmp_xattr_ierrorator::future<> PGBackend::cmp_xattr(
+  const ObjectState& os,
+  OSDOp& osd_op,
+  object_stat_sum_t& delta_stats) const
+{
+  std::string name{"_"};
+  auto bp = osd_op.indata.cbegin();
+  bp.copy(osd_op.op.xattr.name_len, name);
+
+  logger().debug("cmpxattr on obj={} for attr={}", os.oi.soid, name);
+  return getxattr(os.oi.soid, std::move(name)).safe_then_interruptible(
+    [&delta_stats, &osd_op] (auto &&xattr) -> cmp_xattr_ierrorator::future<> {
+    return do_cmp_xattr_mode(osd_op, std::move(xattr)).safe_then_interruptible(
```

Comment:
do_cmp_xattr_mode doesn't do any IO, right?  Can it simply return an error code?